### PR TITLE
refactor: Change mechanism of building encryption hash

### PR DESCRIPTION
**Pull request by pmatyjasek-sumo**
Change mechanism of building encryption hash. Use combination of collector name, access ID and access key.

Signed-off-by: Patryk Matyjasek <pmatyjasek@sumologic.com>

### CODE REVIEW

File: pkg/extension/sumologicextension/README.md [hunk 1]
```diff
@@ -82,3 +82,13 @@ service:
...
+
+## Storing credentials
+
+When collector is starting first time Sumo Logic extension is using `access_key` and `access_id` to register collector with API.
+After registration extension gets collector credentials which are used to authenticate instance of collector.
+Credentials are stored on filesystem in `collector_credentials_path` which is by default set to `$HOME/.sumologic-otel-collector`.
+Name of file that contains credentials is created by hashing combination of `collector_name`, `access_id` and `access_key`.
+This mechanism allows to keep a state of collector (whether is registered or not).
+When collector is restarting it checks if state file exists in `collector_credentials_path`.
+If user want to register collector on the same machine and use another state file, need add the collector name to otc config.
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/README.md:94]:
```suggestion
When collector is starting for the first time, Sumo Logic extension is using `access_key` and `access_id`
to register collector with API.
Upon registration, the extension gets collector credentials which are used to authenticate the collector
when sending request to API (heartbeats, sending data etc).
Credentials are stored on local filesystem to be reused when collector gets restarted (to prevent re-registration).a
The path that's used to store the credentials files is configured via `collector_credentials_path` which is by default
set to `$HOME/.sumologic-otel-collector`.
Name of that file that contains the credentials is created by hashing a string being a combination of `collector_name`, `access_id` and `access_key`.
This mechanism allows to keep the state of the collector (whether it is registered or not).
When collector is restarting it checks if the state file exists in `collector_credentials_path`.
If one would like to register another collector on the same machine then `collector_name` configuration property
has to be specified in order to register the collector under that specific name which will be used to create
a separate state file.
```

**Review by pmalek-sumo — COMMENTED** [review]
There's one more thing we need to add in this PR which is to add a default prefix to the collector name when an empty name (default) is being specified.

Something like `sumologic-otel-collector` (and then the following registrations with the same name would create collectors with this name+timestamp suffix.

File: pkg/extension/sumologicextension/credentials.go [hunk 2]
```diff
@@ -48,7 +48,8 @@ type credsGetter struct {
 // CheckCollectorCredentials checks if collector credentials can be found in path
 // configured in the config.
 func (cr credsGetter) CheckCollectorCredentials() bool {
-	filenameHash, err := hash(cr.conf.CollectorName)
+	key := cr.conf.CollectorName + cr.conf.Credentials.AccessID + cr.conf.Credentials.AccessKey
+	filenameHash, err := hash(key)
 	if err != nil {
```
Comment (astencel-sumo) [inline pkg/extension/sumologicextension/credentials.go:53]: Sorry, this is not strictly related to this PR, but shouldn't we be logging these errors?

File: pkg/extension/sumologicextension/extension.go [hunk 3]
```diff
@@ -124,7 +124,8 @@ func (se *SumologicExtension) storeCollectorCredentials() error {
 	if err := ensureStoreCredentialsDir(se.conf.CollectorCredentialsPath); err != nil {
 		return err
 	}
-	filenameHash, err := hash(se.conf.CollectorName)
+	key := se.conf.CollectorName + se.conf.Credentials.AccessID + se.conf.Credentials.AccessKey
```
Comment (astencel-sumo) [inline pkg/extension/sumologicextension/extension.go:127]: Shouldn't the `storeCollectorCredentials` and `ensureStoreCredentialsDir` functions be part of the `credsGetter`?

This would also make it easier to encapsulate the `CollectorName + AccessID + AccessKey` hash creation, which is repeated here and in `credsGetter.GetStoredCredentials` function.

Comment (astencel-sumo) [reply pkg/extension/sumologicextension/extension.go:127]: Probably a better name than `credsGetter` would then be `credsStore` because it would serve for both retrieving and storing the credentials.

Comment (pmalek-sumo) [reply pkg/extension/sumologicextension/extension.go:127]: I like `credentialsStore` and I agree that it might be more obvious what it does granted we also have a method to store the credentials on it (not only to get them). 👍 

As for repeated hashing definition: I'd even go as far as make a function that would take the config and return the hashed string. This way if we ever decide to change the implementation this will be the only place to change. 👍 

@pmatyjasek-sumo WDYT?

As for other reorganization ideas I'm in for it but I'd suggest to skip the discussion on those in this PR

Comment (astencel-sumo) [reply pkg/extension/sumologicextension/extension.go:127]: > make a function that would take the config and return the hashed string
💯

Comment (pmatyjasek-sumo) [reply pkg/extension/sumologicextension/extension.go:127]: > Probably a better name than `credsGetter` would then be `credsStore` because it would serve for both retrieving and storing the credentials.

Thats why those `storeCollectorCredentials` and `ensureStoreCredentialsDir` are not parts of `credsGetter`. It grabs the authentication data for collector in two ways. If collector wasn't previously registered (file does not exists on filesystem) it register collector through the API and "gets" credentials. If file exists it also "gets" credentials but in this scenario it is taken from file on filesystem. Also both methods requires the same data as `AccessID`, `AccessKey` and `CollectorName` which are stored in config. `storeCollectorCredentials` needs some additional data like `collector credentials`

File: pkg/extension/sumologicextension/extension.go [hunk 4]
```diff
@@ -57,20 +60,30 @@ const (
...
 func newSumologicExtension(conf *Config, logger *zap.Logger) (*SumologicExtension, error) {
+	collectorName := ""
+	uuid := uuid.New()
+	hostname, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
 	if conf.CollectorName == "" {
-		return nil, errors.New("collector name is unset")
+		conf.CollectorName = fmt.Sprintf("%s-%s", hostname, uuid.String())
```
Comment (sumo-drosiek) [inline pkg/extension/sumologicextension/extension.go:70]: If I restart collector, is a new one created, or the previous name is being used?

Comment (pmalek-sumo) [reply pkg/extension/sumologicextension/extension.go:70]: @pmatyjasek-sumo is working on this. The idea is that when you don't specify the collector name (i.e. `""`) you'll get one specific name (based on hostname as far as I recall correctly) and yes it should be the same on each subsequent run with the same empty collector name.

**Comment by sumo-drosiek** [comment]
One last concern:

If you don't set `clobber` and restart collector multiple times, you will get:
`collector-name-<timestamp1>-<timestamp2>-<timestamp3>-<timestamp4>`

Am I right?

File: pkg/extension/sumologicextension/extension.go [hunk 5]
```diff
@@ -57,40 +61,60 @@ const (
...
+	if err != nil {
+		return nil, err
+	}
+	credentialsGetter := credsGetter{
+		conf:   conf,
+		logger: logger,
+	}
 	if conf.CollectorName == "" {
-		return nil, errors.New("collector name is unset")
+		key := fmt.Sprintf("%s%s%s", conf.CollectorName, conf.Credentials.AccessID, conf.Credentials.AccessKey)
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension.go:75]: extract this "algorithm" of creating a key to a function (receiving a config?)

File: pkg/extension/sumologicextension/extension.go [hunk 6]
```diff
@@ -57,40 +61,60 @@ const (
...
+		// If collector name is not set by the user check if the collector was restarted
+		if !credentialsGetter.CheckCollectorCredentials(key) {
+			// If credentials file is not stored on filesystem generate collector name
+			collectorName = fmt.Sprintf("%s-%s", hostname, uuid.String())
+		}
+	} else {
+		collectorName = conf.CollectorName
 	}
 	if conf.Credentials.AccessID == "" || conf.Credentials.AccessKey == "" {
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension.go:84]: This can be checked as first thing in the function, fast return on empty keys

File: pkg/extension/sumologicextension/extension.go [hunk 7]
```diff
@@ -30,17 +30,21 @@ import (
...
 
+	"github.com/google/uuid"
+
 	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/sumologicextension/api"
 	"go.opentelemetry.io/collector/component"
 	"go.uber.org/zap"
 )
 
 type SumologicExtension struct {
+	collectorName     string
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension.go:41]: As suggested in the `pkg/extension/sumologicextension/api/register.go`: let's not add this here as it's not obvious what it's being used for: we already have the config? Why add another `collectorName` in the extension struct itself? (those might be the questions that someone reading this for the first time might be asking)

File: pkg/extension/sumologicextension/extension.go [hunk 8]
```diff
@@ -57,40 +61,60 @@ const (
...
+	} else {
+		collectorName = conf.CollectorName
 	}
 	if conf.Credentials.AccessID == "" || conf.Credentials.AccessKey == "" {
 		return nil, errors.New("access_key and/or access_id not provided")
 	}
 	if conf.HeartBeatInterval <= 0 {
 		conf.HeartBeatInterval = DefaultHeartbeatInterval
 	}
+	hashKey := fmt.Sprintf("%s%s%s", conf.CollectorName, conf.Credentials.AccessID, conf.Credentials.AccessKey)
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension.go:90]: as commented elsewhere: this should be a func

File: pkg/extension/sumologicextension/extension.go [hunk 9]
```diff
@@ -57,40 +61,60 @@ const (
...
+		logger: logger,
+	}
 	if conf.CollectorName == "" {
-		return nil, errors.New("collector name is unset")
+		key := fmt.Sprintf("%s%s%s", conf.CollectorName, conf.Credentials.AccessID, conf.Credentials.AccessKey)
+		// If collector name is not set by the user check if the collector was restarted
+		if !credentialsGetter.CheckCollectorCredentials(key) {
+			// If credentials file is not stored on filesystem generate collector name
+			collectorName = fmt.Sprintf("%s-%s", hostname, uuid.String())
+		}
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension.go:80]: I believe we need to account for more facts here: if user specifies `clobber` as `true` then do we really want to append uuid? Doesn't make sense to me, the user specified by the clobber flag: 

> Please override that collector name for me, whatever it is. (if it's empty, override the default name that corresponds to empty collector name in the config)

File: pkg/extension/sumologicextension/extension.go [hunk 10]
```diff
@@ -148,7 +172,7 @@ func (se *SumologicExtension) storeCollectorCredentials(credentials api.OpenRegi
 		return err
 	}
 
-	encrypedCreds, err := encrypt(collectorCreds, se.conf.CollectorName)
+	encrypedCreds, err := encrypt(collectorCreds, se.hashKey)
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension.go:175]:
```suggestion
	encryptedCreds, err := encrypt(collectorCreds, se.hashKey)
```

File: pkg/extension/sumologicextension/extension.go [hunk 11]
```diff
@@ -57,40 +61,60 @@ const (
 )
 
 func newSumologicExtension(conf *Config, logger *zap.Logger) (*SumologicExtension, error) {
+	collectorName := ""
+	uuid := uuid.New()
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension.go:65]: This is just used once, just use it in place

File: pkg/extension/sumologicextension/extension.go [hunk 12]
```diff
@@ -30,17 +30,21 @@ import (
 	"sync"
 	"time"
 
+	"github.com/google/uuid"
+
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension.go:34]:
```suggestion
	"github.com/google/uuid"
```

**Review by pmalek-sumo — CHANGES_REQUESTED** [review]

File: pkg/extension/sumologicextension/extension.go (hunk 7 above)
Comment (pmatyjasek-sumo) [reply pkg/extension/sumologicextension/extension.go:41]: This is added to avoid overriding user configuration.

**Comment by pmalek-sumo** [comment]
Sure, but this is not obvious what it will be used for (especially since
it's the first field in the struct now)

File: pkg/extension/sumologicextension/extension.go [hunk 13]
```diff
@@ -57,50 +61,85 @@ const (
...
+		conf:   conf,
+		logger: logger,
+	}
 	if conf.CollectorName == "" {
-		return nil, errors.New("collector name is unset")
+		key := createHashKey(conf)
+		// If collector name is not set by the user check if the collector was restarted
+		if !credentialsGetter.CheckCollectorCredentials(key) {
+			// If credentials file is not stored on filesystem generate collector name
+			collectorName = fmt.Sprintf("%s-%s", hostname, uuid.New().String())
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension.go:78]:
```suggestion
			collectorName = fmt.Sprintf("%s-%s", hostname, uuid.New())
```

If a struct implements [`Stringer`](https://golang.org/pkg/fmt/#Stringer) then no need to add `.String()` call in format string for `%s`

File: pkg/extension/sumologicextension/extension.go [hunk 14]
```diff
@@ -57,50 +61,85 @@ const (
...
-		colCreds, err = se.credentialsGetter.GetStoredCredentials()
+	if se.credentialsGetter.CheckCollectorCredentials(se.hashKey) {
+		colCreds, err = se.credentialsGetter.GetStoredCredentials(se.hashKey)
 		if err != nil {
 			return err
 		}
-		se.logger.Info("Found stored credentials")
+		colName := colCreds.CollectorName
+		if !se.conf.Clobber {
+			if se.collectorName == "" {
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension.go:116]: Can we add a comment when this happens?

File: pkg/extension/sumologicextension/extension.go [hunk 15]
```diff
@@ -57,50 +61,85 @@ const (
...
+				se.collectorName = colName
+			}
+			se.logger.Info("Found stored credentials")
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension.go:119]:
```suggestion
			se.logger.Info("Found stored credentials, skipping registration",
				zap.String("collector_name", se.collectorName),
			)
```

**Review by pmalek-sumo — COMMENTED** [review]
Can we please add more tests for those added use cases (especially with empty collector name and `clobber` flag)?

**Review by pmalek-sumo — COMMENTED** [review]
What about the tests 😬  ? :)

File: pkg/extension/sumologicextension/extension_test.go [hunk 16]
```diff
@@ -223,3 +226,88 @@ func TestStoreCredentials(t *testing.T) {
...
+	cfg.CollectorCredentialsPath = dir
+
+	se, err := newSumologicExtension(cfg, zap.NewNop())
+	require.NoError(t, err)
+	require.NoError(t, se.Start(context.Background(), componenttest.NewNopHost()))
+	assert.NotEmpty(t, se.collectorName)
+	t.Cleanup(func() {
+		srv.Close()
+		os.RemoveAll(dir)
+	})
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension_test.go:265]: Move higher up

File: pkg/extension/sumologicextension/extension_test.go [hunk 17]
```diff
@@ -223,3 +226,88 @@ func TestStoreCredentials(t *testing.T) {
...
+	cfg.ExtensionSettings = config.ExtensionSettings{}
+	cfg.ApiBaseUrl = srv.URL
+	cfg.Credentials.AccessID = "dummy_access_id"
+	cfg.Credentials.AccessKey = "dummy_access_key"
+	cfg.CollectorCredentialsPath = dir
+
+	se, err := newSumologicExtension(cfg, zap.NewNop())
+	require.NoError(t, err)
+	require.NoError(t, se.Start(context.Background(), componenttest.NewNopHost()))
+	assert.NotEmpty(t, se.collectorName)
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension_test.go:261]: add TODO to narrow down the checks

File: pkg/extension/sumologicextension/extension_test.go [hunk 18]
```diff
@@ -223,3 +226,88 @@ func TestStoreCredentials(t *testing.T) {
 		require.NoError(t, se.Start(context.Background(), componenttest.NewNopHost()))
 	})
 }
+
+func TestEmptyCollectorName(t *testing.T) {
```
Comment (pmalek-sumo) [inline pkg/extension/sumologicextension/extension_test.go:230]:
```suggestion
func TestRegister_EmptyCollectorName(t *testing.T) {
```

**Review by pmalek-sumo — APPROVED** [review]
